presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java

```diff
@@ -65,6 +65,7 @@ public void testDefaults()
                 .setWriterSortBufferSize(new DataSize(64, Unit.MEGABYTE))
                 .setForceLocalScheduling(false)
                 .setMaxConcurrentFileRenames(20)
+                .setMaxConcurrentFileCreates(100)
```

Comment:
I'm curious how you chose the default? Why is it 5x more than renames? Is it because threads sit idle longer waiting for the file creation to happen? Also, have you tested whether this creates extra load on the coordinator?

Reply:
> I'm curious how you chose the default. Why is it 5x more than renames? Is it because threads sit idle longer waiting for the file creation to happen?

Effectively the first one is used only when commiting a single bucket. The number of files that have to be renamed scales with the number of partitions in the destination table. So it looks like 20 threads should be enough to be able to perform this operation in a reasonable time.

However when inserting into a table that has empty buckets - number of empty files to be created scales as `number of buckets  * number of partitions`, thus the higher number of threads is required to be able to finish it in a reasonable time.

> Also, have you tested whether this creates extra load on the coordinator?

I would expect the threads to be mostly doing IO. We only spend CPU on materializing 0 rows file, what in theory shouldn't be CPU consuming. However that could be file format specific. I think there few things we can do here:

1. Decrease number of threads to something safe (say, 20? or even 10 or 5). 
2. Optimize empty file creating code. Instead of materializing zero-row file content for every file it can be materialized once, and then written in multiple threads to the destination file system. That will make sure that threads will be mostly waiting.
3. Do the #1 first, and then create a separate PR that addresses #2

@lnyng, what do you feel?

Reply:
Thanks for the comment. The number is based on that the current default for empty bucket file creation is ~4k, and @arhimondr suggests that the creation latency for each file could be ~50ms. Setting default to 100 reduces the latency from ~3min to ~2s. If using only 20 threads, the latency would be ~10s or more with higher latency or more bucket files. I think 2s latency is acceptable so I pick 100.

When I was testing a single node hive is used and the creation latency is ~500us so the benefit is not significant. But I also manually adding latency to the file creation logic and this new feature works as expected.

For the extra load I was not considering it yet. What metric would you recommend for measuring that?

Reply:
> For the extra load I was not considering it yet. What metric would you recommend for measuring that?

I think the concern here is that the creating zero rows files might not be just waiting on IO, but actually actively spending CPU cycles. If that's the case - having 100 threads doing it simultaneously may cause some undesired hiccups on coordinator. 

You can try to create multiple zero-rows files in a loop using different storage formats (e.g.: ORC, PARQUET, RCBINARY, etc.) with a reasonable sized table schema (say ~100 different columns). Then you can measure how much CPU time does it take using the operation system cpu counters (`ThreadMXBean#getCurrentThreadCpuTime`) and compare it to the wall time (`System#nanoTime`). If the fraction of the CPU time is not negligible - i would suggest doing the optimization i explained. But again, it doesn't feel like it has to be in this PR. Thus it is up to you.